### ISSUE TITLE
Pyinstaller unable to package correctly some dependencies on Linux

### ISSUE THREAD

**Issue by JohnFarl** [pyinstaller/pyinstaller#6085]
I have tried to export a script with pyinstaller on Linux reading the warnings I have fixed many issues adding hidden imports in spec file but despite I have added hidden imports for all the warning reported in logs with
tf_hidden_imports += collect_submodules('imodulename')

When I execute it crashes with 
```
Unknown Resource Path: <class importlib_resources_adapters.CompatibilityFiles.SpecPath>;
<importlib_resources_adapters.CompatibilityFiles.SpecPath.object at 0x7ff7b945>;
```

The error happens at this line `import from tflite_model_maker import configs`
I have spent 3 days trying to fix this error with no result. 
There is something seriously wrong in how pyinstaller packages dependencies.

pyinstaller 4.4

**Comment by rokm** [comment]
What version of python? And what version of packages involved (or rather, what's the output of `pip freeze˙?).

I can freeze a simple script containing
```from tflite_model_maker import configs```
and it starts without errors, once `tensorflow`'s source (.py) files are explicitly collected? (But even prior to that, the error was explicit `OSError: could not get source code`).

**Comment by JohnFarl** [comment]
@rokm Python 3.8 on Ubuntu x64

The output of pip freeze is 
```
absl-py==0.10.0
altgraph==0.17
appdirs==1.4.4
apt-xapian-index==0.49
apturl==0.5.2
astunparse==1.6.3
attrs==21.2.0
audioread==2.1.9
beautifulsoup4==4.9.1
blinker==1.4
cachetools==4.1.1
certifi==2020.6.20
cffi==1.14.2
chardet==3.0.4
click==7.1.2
colorama==0.4.3
command-not-found==0.3
cryptography==3.1
cupshelpers==1.0
cycler==0.10.0
Cython==0.29.21
dataclasses==0.6
dbus-python==1.2.16
decorator==5.0.9
dill==0.3.4
distro==1.5.0
distro-info===0.23ubuntu1
dm-tree==0.1.6
empy==3.3.4
entrypoints==0.3
fire==0.4.0
flatbuffers==1.12
future==0.18.2
gast==0.3.3
genmsg==0.5.14
genpy==0.6.9
gin-config==0.4.0
google-api-core==1.31.1
google-api-python-client==2.15.0
google-auth==1.21.0
google-auth-httplib2==0.1.0
google-auth-oauthlib==0.4.1
google-cloud-bigquery==2.23.2
google-cloud-core==1.7.2
google-crc32c==1.1.2
google-pasta==0.2.0
google-resumable-media==1.3.3
googleapis-common-protos==1.53.0
gpg===1.13.1-unknown
grpcio==1.32.0
h5py==2.10.0
html5lib==1.1
httplib2==0.18.1
idna==2.10
importlib-resources==5.2.2
jeepney==0.4.3
joblib==1.0.1
kaggle==1.5.12
keras-nightly==2.5.0.dev2021032900
Keras-Preprocessing==1.1.2
keyring==21.4.0
kiwisolver==1.3.1
language-selector==0.1
launchpadlib==1.10.13
lazr.restfulclient==0.14.3
lazr.uri==1.0.5
librosa==0.8.1
llvmlite==0.36.0
lxml==4.6.3
Markdown==3.2.2
matplotlib==3.4.2
netifaces==0.10.9
neural-structured-learning==1.3.1
numba==0.53.1
numpy==1.19.5
oauthlib==3.1.0
olefile==0.46
opencv-python-headless==4.5.3.56
opt-einsum==3.3.0
packaging==20.4
pandas==1.3.1
pbr==5.4.5
pexpect==4.8.0
Pillow==7.2.0
pkgconfig==1.5.1
ply==3.11
pooch==1.4.0
promise==2.3
proto-plus==1.19.0
protobuf==3.13.0
psutil==5.8.0
ptyprocess==0.6.0
py-cpuinfo==8.0.0
pyasn1==0.4.8
pyasn1-modules==0.2.8
pybind11==2.7.0
pycairo==1.16.2
pycocotools==2.0.2
pycparser==2.20
pycrypto==2.6.1
pycups==1.9.73
Pygments==2.6.1
PyGObject==3.36.0
pyinstaller==4.4
pyinstaller-hooks-contrib==2021.2
PyJWT==1.7.1
pyparsing==2.4.7
PyQt5==5.15.0
PyQt5-sip==12.8.1
python-apt==2.0.0+ubuntu0.20.4.5
python-dateutil==2.8.2
python-debian==0.1.37
python-slugify==5.0.2
pytz==2021.1
PyYAML==5.3.1
reportlab==3.5.48
requests==2.24.0
requests-oauthlib==1.3.0
requests-unixsocket==0.2.0
resampy==0.2.2
rsa==4.6
scikit-learn==0.24.2
scipy==1.5.2
SecretStorage==3.1.2
sentencepiece==0.1.96
simplejson==3.17.2
sip==5.4.0
six==1.15.0
SoundFile==0.10.3.post1
soupsieve==2.0.1
systemd-python==234
tensorboard==2.5.0
tensorboard-data-server==0.6.1
tensorboard-plugin-wit==1.7.0
tensorflow==2.4.1
tensorflow-addons==0.13.0
tensorflow-datasets==4.4.0
tensorflow-estimator==2.4.0
tensorflow-hub==0.11.0
tensorflow-metadata==1.2.0
tensorflow-model-optimization==0.6.0
tensorflowjs==3.8.0
termcolor==1.1.0
testresources==2.0.1
text-unidecode==1.3
tf-models-official==2.3.0
tf-slim==1.1.0
tflite-model-maker==0.2.5
tflite-support==0.1.0rc4
threadpoolctl==2.2.0
tinyaes==1.0.1
toml==0.10.1
tqdm==4.62.0
typeguard==2.12.1
typing-extensions==3.7.4.3
ubuntu-advantage-tools==27.2
ubuntu-drivers-common==0.0.0
ufw==0.36
unattended-upgrades==0.1
uritemplate==3.0.1
urllib3==1.25.10
usb-creator==0.3.7
wadllib==1.3.4
webencodings==0.5.1
Werkzeug==1.0.1
wrapt==1.12.1
xkit==0.0.0
zipp==3.5.0
```

**Comment by rokm** [comment]
Ah, this looks like lack of proper support for `importlib_resources` / `importlib.resources` in PyInstaller, which makes it a duplicate of #5616. 

That said, you can work around it by adding `--debug noarchive` to your PyInstaller command (if building directly from .py file) or setting `noarchive=True` argument to `EXE()˙ in the spec file. This will cause collected .pyc modules to be stored directly in dist directory instead of in executable-embedded archive, which in turn allows existing importlib facilities to deal with them.

PS: you should also remove `dataclasses` from your environment. It is probably listed as dependency by one of other packages, but it is unnecessary with python 3.7 and later, and is actually causing issues that result in `pandas` modules not being properly collected.

**Comment by rokm** [comment]
For the record, the reported error comes from [this part of `tensorflow_datasets`](https://github.com/tensorflow/datasets/blob/master/tensorflow_datasets/core/utils/resource_utils.py#L81-L97). In the frozen application, `importlib_resources.files('tensorflow_datasets')` returns an `importlib_resources_adapters.CompatibilityFiles.SpecPath` as a compatibility object for accessing resources (due to lack of resource reader implementation in PyInstaller's FrozenImporter), which is incompatible with assumptions made by `tensorflow_datasets`. This "fallback" SpecPath is used because in a frozen application, the packages' modules are not found on disk (they are in embedded archive); if `noarchive` option is used, the adapter for on-disk packages is used instead, and its `files()` implementation returns `pathlib.Path` instance.

The issue should be fixed by #6195 as our implementation's `files()` method returns pathlib.Path instance (at least for the time being).
